source/azure_iot.c

```diff
@@ -0,0 +1,410 @@
+    {
+        AZLogError( ( "AzureIoTMessagePropertiesFind failed: Invalid argument \r\n" ) );
+        return eAzureIoTInvalidArgument;
+    }
+
+    xResult = az_iot_message_properties_find( &pxMessageProperties->_internal.xProperties,
+                                             xNameSpan, &outValueSpan );
+
+    if( az_result_failed( xResult ) )
+    {
+        return eAzureIoTItemNotFound;
+    }
+
+    *ppucOutValue = az_span_ptr( outValueSpan );
+    *pulOutValueLength = ( uint32_t ) az_span_size( outValueSpan );
+
+    return eAzureIoTSuccess;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTResult_t AzureIoT_Base64HMACCalculate( AzureIoTGetHMACFunc_t xAzureIoTHMACFunction,
+                                              const uint8_t * pucKey,
+                                              uint32_t ulKeySize,
+                                              const uint8_t * pucMessage,
+                                              uint32_t ulMessageSize,
+                                              uint8_t * pucBuffer,
+                                              uint32_t ulBufferLength,
+                                              uint8_t * pucOutput,
+                                              uint32_t ulOutputSize,
+                                              uint32_t * pulOutputLength )
+{
+    AzureIoTResult_t xStatus;
+    uint8_t * pucHashBuf;
+    uint32_t ulHashBufSize = 33;
+    uint32_t ulBinaryKeyBufSize;
+
+    if( ( xAzureIoTHMACFunction == NULL ) ||
+        ( pucKey == NULL ) || ( ulKeySize == 0 ) ||
+        ( pucMessage == NULL ) || ( ulMessageSize == 0 ) ||
+        ( pucBuffer == NULL ) || ( ulBufferLength == 0 ) ||
+        ( pucOutput == NULL ) || ( pulOutputLength == NULL ) )
+    {
+        AZLogError( ( "AzureIoT_Base64HMACCalculate failed: Invalid argument \r\n" ) );
+        return eAzureIoTInvalidArgument;
+    }
+
+    ulBinaryKeyBufSize = ulBufferLength;
+    xStatus = prvAzureIoTBase64Decode( ( char * ) pucKey, ulKeySize,
+                                      pucBuffer, ulBinaryKeyBufSize, &ulBinaryKeyBufSize );
+
+    if( xStatus )
+    {
+        return xStatus;
+    }
+
+    ulBufferLength -= ulBinaryKeyBufSize;
+
+    if( ulHashBufSize > ulBufferLength )
+    {
+        return eAzureIoTOutOfMemory;
+    }
+
+    pucHashBuf = pucBuffer + ulBinaryKeyBufSize;
+    memset( pucHashBuf, 0, ulHashBufSize );
+
+    if( xAzureIoTHMACFunction( pucBuffer, ulBinaryKeyBufSize,
+                               pucMessage, ( uint32_t ) ulMessageSize,
+                               pucHashBuf, ulHashBufSize, &ulHashBufSize ) )
+    {
+        return eAzureIoTFailed;
+    }
+
+    xStatus = prvAzureIoTBase64Encode( pucHashBuf, ulHashBufSize,
+                                      ( char * ) pucOutput, ulOutputSize );
+
+    if( xStatus )
+    {
+        return xStatus;
+    }
+
+    *pulOutputLength = ( uint32_t ) strlen( ( char * )pucOutput );
```

Comment:
can we return length from prvAzureIoTBase64Encode so that we do not have to link strlen?

Reply:
yea good call